### ISSUE TITLE
[Feature Request] Package custom Android resources 

### ISSUE THREAD

**Issue by marchbold** [Gamua/Adobe-Runtime-Support#163]
## Problem Description

Currently in order to add custom resources for an Android application we have to package an ANE containing the resources.

This is often required for configuration of SDKs or additional assets (icons/sounds). Most commonly:
- Notification icons for local and push notifications (Firebase/OneSignal etc)
- Custom sounds for notifications
- Firebase configuration values.xml

## Known Workarounds

We have a script to build an ANE:

https://github.com/distriqt/ANE-CustomResources

You add your resources into this ANE and then package the ANE with your application.

Alternatively you can add your resources into the AIR SDK however this requires modifying the SDK for every application build.

## Request

Would be much easier if we could provide a "res" folder (could be specified in the application descriptor perhaps?) when packaging an Android application with AIR and AIR could merge these resources with the application.

**Comment by Mintonist** [comment]
It will be very useful!

**Comment by rewb0rn** [comment]
Yes, would be nice to have this! There is another way to achieve this without an ANE actually. You can copy your files into the `[Air SDK}/lib/android/lib/resources/app_entry/res` and then they will be included in the app. But it becomes tricky as soon as you use several different applications that need different sets of resources. Would be great to have this simplified with a res folder in the application.xml as suggested.

**Comment by Oldes** [comment]
@rewb0rn that was mentioned as a workaround:

> Alternatively you can add your resources into the AIR SDK however this requires modifying the SDK for every application build.

**Comment by rewb0rn** [comment]
Yes, you are right, my bad.

**Comment by ajwfrost** [comment]
@marchbold I'd missed this one .. looks like this should be relatively straightforward, so then we have two options:
1) add a new setting into the AIR XML descriptor file, or
2) add a new parameter to the ADT packaging command

My preference would be for (2) but I'm mindful that some IDEs may not let you add custom parameters.. 

Any thoughts?

**Comment by marchbold** [comment]
Honestly I was just thinking something simple like if the developer packaged a specially named directory e.g `res` then the packager would treat this as android resources and include them in the build.  Similar to the `Frameworks` directory for iOS? 

Maybe you could include an option to configure this name but I think this is important enough to just be done in a specified directory.

**Comment by myflashlab** [comment]
Yep, Very much interested in a solution similar to what @marchbold suggested above.

**Comment by ajwfrost** [comment]
Okay thanks guys - so if we spot a "res" folder in the package contents for an Android build, we can treat those as resources and merge them as if someone was using the manual hack mentioned as the second workaround..

**Comment by marchbold** [comment]
@ajwfrost That is all we need and will honestly be a game changer!  :)

**Comment by myflashlab** [comment]
Awesome, this will make Firebase and FCM implementations a lot easier. Yoo-hoo.

**Comment by marchbold** [comment]
@ajwfrost Is this available in the latest build?! Super excited!

**Comment by ajwfrost** [comment]
Yes, with a slight caveat that we've not extensively tested this! It should be compiling the resources properly and worked for our basic tests, but I'm aware people are always doing things in bigger and different ways from what we expect! So, feedback is welcome :-)

**Comment by Michoko92** [comment]
Excellent! Does it mean that now we can add adaptive icons easily?

**Comment by leossmith** [comment]
Wow! This will be so much faster to work with. 

> Excellent! Does it mean that now we can add adaptive icons easily?

As adaptive icons are just some files in Res, this will allow you to just copy paste the needed files and it should work.

**Comment by marchbold** [comment]
@Michoko92 This should simplify any process where you had to package a custom ANE or modify the AIR SDK resources, including:

- adaptive icons
- notification icons
- notification sounds
- Firebase / Google Services configs

**Comment by Michoko92** [comment]
I made some tests, and adaptive icons seem to work yes! Just be sure you remove the same files in "<YOUR_AIR_SDK_PATH>\lib\android\lib\resources\app_entry\res" folder, or they will take precedence over the ones you have in your local "res" folder.  

Awesome job, Harman team!

**Comment by ajwfrost** [comment]
@Michoko92  thanks, sounds good .. we did debate about the priorities here and whether the user-specified resources should override the SDK ones. I'd be interested if folk think that the user-specified ones should have the precedence, I was slightly concerned that we may lose some resources that we depend upon somewhere so told the developer to add in the users resources last...

**Comment by rewb0rn** [comment]
I would prefer to have user-specified resources overwriting the Air-SDK resources. This way, anything coming from Air would be "default" resources with the user being able to choose whether to overwrite them or not.

**Comment by Michoko92** [comment]
> I would prefer to have user-specified resources overwriting the Air-SDK resources. This way, anything coming from Air would be "default" resources with the user being able to choose whether to overwrite them or not.

I think it might be better, yes. No big deal, but having the SDK only provide the resources we didn’t override locally should make more sense.

**Comment by leossmith** [comment]
Indeed. It would be better to have a local folder overriding the SDK. If a project has it, everything works there, if not (resist the urge to use else!) the SDK one is used.

**Comment by ajwfrost** [comment]
Okay :-) I get the feeling it's going to be unanimous! Simple fix (I think) so we'll switch it round for the next release..
Thanks

**Comment by megajogos** [comment]
We just updated our SDK to AIR 33.1.1.259 and we got this error:

```
ADT error:
aapt tool failed:invalid resource directory name: C:\Users\fontec\AppData\Local\Temp\00172c11-aa20-4910-bb4f-1746b9616466\user_provided_res avBot10.png
```

We already have an folder called "res" but it is not an android res folder. Can I disable this new feature?

**Comment by ajwfrost** [comment]
Ah, I was slightly concerned about this. So if you were looking to just have a folder called 'res' that resulted in an APK with a subfolder in /assets/res, this isn't going to work with the new set-up...

A few options we could look at:
1) add a new switch to ADT to disable this new feature
2) have the new feature disabled by default and require a command-line switch in ADT to turn it on..
3) check for the existence of a file within a 'res' folder to decide how to interpret/package it
Any other suggestions?

**Comment by Oldes** [comment]
@ajwfrost I vote for `2`, so to have a switch with the path to the folder.. it could be called `-resdir` as there is already `-extdir`

**Comment by Oldes** [comment]
Btw.. it would be good to have its behaviour well documented, where there would be described, how it deals with ANE's resources as well.

**Comment by leossmith** [comment]
I just tested it in a demo project and just making a res folder in my src folder, worked like a charm. 

I vote for 1

**Comment by Oldes** [comment]
Option `1` does not help people like **megajogos** who are already using `res` dir and still would want to use the feature in a future.

Also I don't know how others, but I'm using single project to produce multiple targets... even multiple Android targets, like for Google Play, Amazon, Tencent.. for each of them there is different manifest file and different need of ANEs, icons etc., so being able to specify path to the right folder is a better than having just one and use folder renaming!

**Comment by Michoko92** [comment]
I'd vote also for a -resdir switch. It is the most flexible solution, and allows to specify an external ressources location without assuming a folder called "res" should systematically be packaged as additional resources.

**Comment by ajwfrost** [comment]
Okay thanks for the feedback..
So for now, if you need the old behaviour, please update your ADT with the below version:
https://transfer.harman.com/messages/6BT0viikyLuSK3qTDsF6af/attachments/uT1n1R1VpqIrLOH8sIlgeo/download/adt.jar
Checksum (SHA256): f59d9338ebc046bdfb38ea18ef303bb482c60926389df169307b6b85d53c40dd

**Comment by megajogos** [comment]
Thanks! 

An option like "-resdir switch" would be great!

**Comment by sjabberwocky** [comment]
I agree that something like "-resdir switch" would be nice, cause we have `res` folder with game resources too.

**Comment by sjabberwocky** [comment]
BTW, Flash Builder does not build APK file with "res" folder using any version of adt.jar file.
Also I noticed that on 259 on AIR Simulator I have about 30 fps instead of 60, which I have on 227.
On device it looks like game works fine.

**Comment by marchbold** [comment]
@ajwfrost Is it hard to auto-detect this? The Android resources are a very specific structure so it shouldn't be hard to determine whether the `res` directory contains android resources or some other game assets. If the res directory matches Android resources, then package it as resources, otherwise just package it as assets. 

I do believe this should be common enough with an Android app to warrant it being the default. There are very few apps that wouldn't need at a minimum notification icons or adaptive app icons.

**Comment by joshtynjala** [comment]
@ajwfrost I'm trying to add support for resdir to vscode-as3mxml, but I can't seem to find an example of the "syntax" for resdir. Can resdir be specified only once? Or can there be multiple resdirs? Should the option include an equals sign or no equals sign before the path (or is it optional)? Thanks!

**Comment by joshtynjala** [comment]
@ajwfrost Also, where exactly does `resdir` need to be included? I know that adt is very, very (very, very!) particular about where most options are located. Does it need to be near the other Android-specific options, like `arch` and `airDownloadURL`? Is it near other file options, like `extdir`?

**Comment by ajwfrost** [comment]
Hi @joshtynjala  - the `-resdir` argument is handled as a file option, similar to `-extdir`. It can only be specified once; no equals sign.
thanks

**Comment by joshtynjala** [comment]
@ajwfrost Thank you!

**Comment by leossmith** [comment]
Any update on when to expect this to be moved to descriptor so it works for intellij?

**Comment by Mintonist** [comment]
Seems it's done in AIR 33.1.1.406!

**Comment by leossmith** [comment]
> Seems it's done in AIR 33.1.1.406!

have you figured out how to use it in the descriptor?

**Comment by marchbold** [comment]
This is now available in 33.1.1.406

I've put together the following guide on it's usage. Let me know if anything isn't clear:

https://docs.airnativeextensions.com/docs/tutorials/android-resources

**Comment by ronyarbiv** [comment]
Hi,

I have opend District-Firebase core- example
<resdir>res</resdir>
but i am getting an error - Element resdir is not allowd here 

using -
intellij - latest version
Air Sdk - Latest (same problem when using .406)
distriqt - Latest

![resdir](https://user-images.githubusercontent.com/4277665/170833808-75c84819-8167-4a08-8ba0-47bbacc4cb76.png)

**Comment by marchbold** [comment]
That just looks like a schema issue in intellij, it's safe to ignore. Hopefully in the next major update to air we'll be able to use a new namespace schema and intellij will pick that up.

**Comment by ajwfrost** [comment]
This may also be because we have missed updating the schema for 33.1 when we've updated the ADT internal validator... so I suspect IntelliJ is looking at the schema/XSD document in order to see whether it's valid.

If you edit the below file:
`templates/air/Descriptor.33.1.xsd`
can you add a line e.g. after the description/copyright/icon lines:
```
        <xs:element name="resdir" type="xs:string" minOccurs="0" />
```

Of course you also need to have the XML descriptor file with a namespace of 33.1 for the new ADT validator, or this schema file, to be picked up..

thanks

**Comment by ronyarbiv** [comment]
Hi,

Thanks for the help.
I have added the line, but i am still getting the error.
Maybe i do something wrong.
Attached the screenshots.
![resdir](https://user-images.githubusercontent.com/4277665/170940501-38c27518-b304-4a99-9084-1c3bcae4d3d5.png)

**Comment by ajwfrost** [comment]
Okay there's something wrong here from what I can see! My line numbering is completely different... there are multiple files called "Descriptor.33.1.xsd" but only one of these has got the word "IconType", which is under the `templates\air` folder. But at line 70-71 I'm seeing the x/y definitions for `application.initialWindow` rather than the type definitions...

In my file, at line 44-46, we have the definitions of descriptor/copyright/icon underneath the application tag. So this would be a good place to add it. See below for my screenshot... can you confirm what SDK version you took this from (including build number and whether it was AIR vs Flex, and which OS) and we can check the templates were set up properly!
 
thanks
![image](https://user-images.githubusercontent.com/6400804/170944310-b2f46f3b-9454-435b-88a6-fd4b5f06c5e2.png)

**Comment by ronyarbiv** [comment]
Thanks
using- 
windows 10
intellij 2022.1.1
AirSdk 33.1.1.856
AIR (no Flex)

I have tried to add the line on 47, attached the screenshot.
It still not recognized. 
Probably i am doing something wrong. 

thank you
![resdir](https://user-images.githubusercontent.com/4277665/170951510-87a2a149-18fa-40fd-bfc5-c4c0a3c67bce.png)

**Comment by ronyarbiv** [comment]
Thanks
Thats an error that prevent it from running and packing.
I have tried to disable inspection for that line but i didn't succeed.

**Comment by ajwfrost** [comment]
Okay .. I hesitate to suggest it, but have you tried restarting IntelliJ - or even the computer - after making those schema edits?

**Comment by ronyarbiv** [comment]
Thanks
Every suggestion is great :)

Sure i was updating it while intellij closed and opend it after it.

Now i have also restarted the coputer, but it still showing the Error.

**Comment by marchbold** [comment]
I think it might be a bit more complex than this. 

I believe the plugin for intellij may actually have prebuilt versions of the xsd in one of the jar files or something. I haven't looked into it in detail but you can see that they have the xsd as part of the source for the plugin here: 

https://github.com/JetBrains/intellij-plugins/tree/master/flex/resources/schemas

So I think the solution will actually be updating the schema there and then rebuilding and updating the plugin

**Comment by ajwfrost** [comment]
Oh dear... I hadn't realised that they were doing this! So yes, until that gets updated, it will be a problem. Short term you could try to just edit the 33.1 schema in their plug-in and rebuild it .. or alternatively, can you switch to using the `resdir` argument on the command line rather than having it in the XML file?

thanks

**Comment by marchbold** [comment]
As a side note, this doesn't stop you from using this in intellij, it's just a warning (albeit an error) that appears in the app-descriptor editor, but you can ignore it and still package and debug. We have been using it without issues.

**Comment by ronyarbiv** [comment]
Thanks.

Can you please tell me how to ignore it in intellij? i have tried to disable inspection for this line but i haven't succeed. it will be much helpful if you might guide me to how to setup intellij for ignoring this error. i was searching on Google and on intellij but didn't find a way. unless i ignore all xml inspections or something like that....

thanks

**Comment by marchbold** [comment]
The error shouldn't stop the build or packaging? Well it doesn't for us.
